modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -1227,6 +1273,379 @@ private void finishCacheStart(UUID reqId, Map<UUID, Boolean> res, Map<UUID, Exce
+
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (log.isInfoEnabled()) {
+            log.info("Starting incremental snapshot restore operation " +
+                "[reqId=" + opCtx0.reqId + ", snapshot=" + opCtx0.snpName + ", incIdx=" + opCtx0.incIdx +
+                ", caches=" + F.viewReadOnly(opCtx0.cfgs, c -> c.config().getName()) + ']');
+        }
+
+        GridFutureAdapter<Boolean> res = new GridFutureAdapter<>();
+
+        ctx.pools().getSnapshotExecutorService().submit(() -> {
+            try {
+                Set<Integer> cacheIds = opCtx0.cfgs.keySet();
+
+                walEnabled(false, cacheIds);
+
+                restoreIncrementalSnapshot(opCtx0.snpName, opCtx0.snpPath, cacheIds, opCtx0.incIdx);
+
+                walEnabled(true, cacheIds);
+
+                CheckpointProgress cp = ctx.cache().context().database()
+                    .forceNewCheckpoint("Incremental snapshot restored.", (fut) -> {
+                        if (fut.error() != null)
+                            res.onDone(fut.error());
+                        else
+                            res.onDone(true);
+                    });
+
+                if (cp == null)
+                    res.onDone(new IgniteCheckedException("Node is stopping."));
+            }
+            catch (Throwable e) {
+                res.onDone(e);
+            }
+        });
+
+        return res;
+    }
+
+    /**
+     * Restore incremental snapshot.
+     *
+     * @param snpName Base snapshot name.
+     * @param snpPath Base snapshot path.
+     * @param cacheIds Restoring cache IDs.
+     * @param incIdx Index of incremental snapshot.
+     */
+    private void restoreIncrementalSnapshot(
+        String snpName,
+        String snpPath,
+        Set<Integer> cacheIds,
+        int incIdx
+    ) throws IgniteCheckedException, IOException {
+        File[] segments = walSegments(snpName, snpPath, incIdx);
+
+        UUID incSnpId = ctx.cache().context().snapshotMgr()
+            .readIncrementalSnapshotMetadata(snpName, snpPath, incIdx)
+            .requestId();
+
+        IncrementalSnapshotFinishRecord incSnpFinRec = readFinishRecord(segments[segments.length - 1], incSnpId);
+
+        if (incSnpFinRec == null)
+            throw new IgniteCheckedException("System WAL records for incremental snapshot wasn't found [id=" + incSnpId + ']');
+
+        CacheStripedExecutor exec = new CacheStripedExecutor(ctx.pools().getStripedExecutorService());
+
+        UUID prevIncSnpId = incIdx > 1
+            ? ctx.cache().context().snapshotMgr().readIncrementalSnapshotMetadata(snpName, snpPath, incIdx - 1).requestId()
+            : null;
+
+        long start = U.currentTimeMillis();
+
+        LongAdder applied = new LongAdder();
+
+        try (WALIterator it = walIter(log, segments)) {
+            IgnitePredicate<GridCacheVersion> txVerFilter = null;
+
+            // Skips applying WAL until base snapshot record.
+            while (it.hasNext()) {
```

Comment:
Looks like we can search for a specific WAL record with the record type filter - `IgniteWalIteratorFactory#IteratorParametersBuilder#addFilter`.
Filter allow do not read whole record to skip it.
So, basically iteration algorithm can be like following:

1. Create `walIterator1` with the filter only `CLUSTER_SNAPSHOT` record.
2. iterate it while find desired record -> snapRec
3. Create `walIterator2` from `WALPointer#next` to iterate further.

Reply:
> Create walIterator2 from WALPointer#next to iterate further.

Unfortunately, it doesn't work for compacted segments this way:
1. WALPointer (that contains offset) is set on original wal segment. 
2. This pointer is read from compacted wal segment. And it consider offset as amount of bytes to skip. But it tries to skip it on compacted segment. As a result more records skipped. 

I suppose we need a separate ticket to fix this behaviour for compacted segment.